### PR TITLE
[COREVM-583] Implement JIT compiler backend based on LLVM's MCJIT framework

### PR DESCRIPTION
**Pull request by yanzhengli**
This patch implements the JIT compiler backend based on LLVM's MCJIT framework.

### CODE REVIEW

File: src/jit/runtime_value_type.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  ValueTypeUInt64,
+  ValueTypeSPF,
+  ValueTypeDPF
+};
+
+struct AggregateType;
+
+typedef corevm::types::variant::variant<ValueType, AggregateType> RuntimeValueType;
+
+struct AggregateType
```
Comment (mizuki-nana) [inline src/jit/runtime_value_type.h]: Add docstring.

File: src/jit/runtime_value_type.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+#include "types/variant/variant.h"
+#include <cstdint>
+#include <vector>
+
+
+namespace corevm {
+namespace jit {
+
+enum class ValueType : uint8_t
```
Comment (mizuki-nana) [inline src/jit/runtime_value_type.h]: Add docstring.

File: tests/jit/llvmmcjit_backend_unittest.cc [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+using namespace llvm;
+using namespace corevm;
+
+// -----------------------------------------------------------------------------
+
+static const char* FUNC_NAME = "fib";
+
+// -----------------------------------------------------------------------------
+
+static Function *CreateFibFunction(Module *M, LLVMContext &Context) {
```
Comment (mizuki-nana) [inline tests/jit/llvmmcjit_backend_unittest.cc]: Maybe update formatting for this function.

File: tests/jit/llvmmcjit_backend_unittest.cc [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  // Create the return instruction and add it to the basic block
+  ReturnInst::Create(Context, Sum, RecurseBB);
+
+  return FibF;
+}
+
+// -----------------------------------------------------------------------------
+
+class LLVMMCJITBackendUnitTest : public ::testing::Test
+{
```
Comment (mizuki-nana) [inline tests/jit/llvmmcjit_backend_unittest.cc]: `{}`

File: tests/jit/llvmmcjit_backend_unittest.cc [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*******************************************************************************/
+#include "jit/jit_compiler_llvmmcjit_backend.h"
+
+#include "llvm/Analysis/Verifier.h"
```
Comment (mizuki-nana) [inline tests/jit/llvmmcjit_backend_unittest.cc]: Probably don't need some of these LLVM headers.

File: tests/jit/llvmmcjit_backend_unittest.cc [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+#include "llvm/IR/DerivedTypes.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Support/TargetSelect.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include <gtest/gtest.h>
+#include <cassert>
+#include <iostream>
```
Comment (mizuki-nana) [inline tests/jit/llvmmcjit_backend_unittest.cc]: Don't need.

File: tests/jit/llvmmcjit_backend_unittest.cc [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+TEST_F(LLVMMCJITBackendUnitTest, TestRunBackend)
+{
+  // Initialize dummy LLVM module.
+  InitializeNativeTarget();
+  InitializeNativeTargetAsmPrinter();
+  InitializeNativeTargetAsmParser();
+  LLVMContext Context;
+  llvm::OwningPtr<llvm::Module> M(new llvm::Module("test", Context));
+  CreateFibFunction(M.get(), Context);
+  assert(!verifyModule(*M));
```
Comment (mizuki-nana) [inline tests/jit/llvmmcjit_backend_unittest.cc]: Either put a comment saying that `verifyModule` returns `0` on success, or just do `assert(0 == verifyModule(*M));`

File: tests/jit/util_unittest.cc [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*******************************************************************************/
+#include "jit/util.h"
+
+#include <gtest/gtest.h>
+
+
+// -----------------------------------------------------------------------------
+
+class UtilUnitTest : public ::testing::Test {};
```
Comment (mizuki-nana) [inline tests/jit/util_unittest.cc]: Rename to `JitUtilUnitTest`

File: tests/jit/llvmmcjit_backend_unittest.cc (hunk 5 above)
Comment (mizuki-nana) [inline tests/jit/llvmmcjit_backend_unittest.cc]: Also they need to be in `<>` instead of double-quotes.

File: tests/jit/util_unittest.cc [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+  const llvm::GenericValue generic_value = jit::convert_to_llvm_generic_value(
+    runtime_value, runtime_value_type);
+
+  ASSERT_EQ(int_val, generic_value.IntVal);
+}
+
+// -----------------------------------------------------------------------------
+
+TEST_F(UtilUnitTest, TestConvertTypeValueToRuntimeValueEquivalent)
```
Comment (mizuki-nana) [inline tests/jit/util_unittest.cc]: `s/TestConvertTypeValueToRuntimeValueEquivalent/TestConvertLLVMTypeValueToRuntimeValueEquivalent/`

File: tests/jit/util_unittest.cc [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+  ASSERT_EQ(aggregate_type.types.size(), generic_value.AggregateVal.size());
+
+  ASSERT_EQ(1, generic_value.AggregateVal[0].IntVal);
+  ASSERT_EQ(3.14, generic_value.AggregateVal[1].DoubleVal);
+}
+
+// -----------------------------------------------------------------------------
+
+TEST_F(UtilUnitTest, TestConvertAggregateTypeValueToRuntimeValueEquivalent)
```
Comment (mizuki-nana) [inline tests/jit/util_unittest.cc]: `s/TestConvertAggregateTypeValueToRuntimeValueEquivalent/TestConvertLLVMAggregateTypeValueToRuntimeValueEquivalent`

File: src/jit/runtime_value_type.h [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  ValueTypeDPF
+};
+
+struct AggregateType;
+
+typedef corevm::types::variant::variant<ValueType, AggregateType> RuntimeValueType;
+
+struct AggregateType
+{
+  std::vector<RuntimeValueType> types;
```
Comment (mizuki-nana) [inline src/jit/runtime_value_type.h]: We should consider use `llvm::SmallVector` here instead.

File: src/jit/jit_compiler_backend.h [hunk 12]
```diff
@@ -62,7 +63,8 @@ class JITCompilerBackend
    * Returns a boolean value indicating whether the evaluation is successful.
    */
   virtual bool eval_func(const std::vector<RuntimeValue>& args,
-    RuntimeValue& result) = 0;
+    const std::vector<RuntimeValueType>& arg_types,
```
Comment (yanzhengli) [inline src/jit/jit_compiler_backend.h]: Done.

File: src/jit/jit_compiler_llvmmcjit_backend.h [hunk 13]
```diff
@@ -56,15 +61,17 @@ class JITCompilerLLVMMCJITBackend : public JITCompilerBackend
    * Implements `JITCompilerBackend::eval_func()`.
    */
   virtual bool eval_func(const std::vector<RuntimeValue>& args,
```
Comment (yanzhengli) [inline src/jit/jit_compiler_llvmmcjit_backend.h]: Done.

**Comment by mizuki-nana** [comment]
```
                                             __----~~~~~~~~~~~------___
                                  .  .   ~~//====......          __--~ ~~
                  -.            \_|//     |||\\  ~~~~~~::::... /~
               ___-==_       _-~o~  \/    |||  \\            _/~~-
       __---~~~.==~||\=_    -_--~/_-~|-   |\\   \\        _/~
   _-~~     .=~    |  \\-_    '-~7  /-   /  ||    \      /
 .~       .~       |   \\ -_    /  /-   /   ||      \   /
/  ____  /         |     \\ ~-_/  /|- _/   .||       \ /
|~~    ~~|--~~~~--_ \     ~==-/   | \~--===~~        .\
         '         ~-|      /|    |-~\~~       __--~~
                     |-~~-_/ |    |   ~\_   _-~            /\
                          /  \     \__   \/~                \__
                      _--~ _/ | .-~~____--~-/                  ~~==.
                     ((->/~   '.|||' -_|    ~~-/ ,              . _||
                                -_     ~\      ~~---l__i__i__i--~~_/
                                _-~-__   ~)  \--______________--~~
                              //.-~~~-~_--~- |-------~~~~~~~~
                                     //.-~~~--\
```

**Pull request merged**